Added file and line to optional assert

alexito4 opened the pull request:
As [suggested in the forums](https://forums.swift.org/t/reducer-optional-assert-with-more-info/38281?u=alejandro_martinez) this PR adds #file and #line default parameters to an `optional()` reducer method to add the information into the assertion message.

Questions I would like to ask:
- Are you happy if the function replaces the property (breaking change) or do you want to keep both?
- Feel free to suggest edits on the wording of the message.
- What other places would you like to see this change? `forEach` maybe?

Once this is clarified I will update the examples and tests 😄

stephencelis commented:
Thanks @alexito4!

> Questions I would like to ask:
> 
> * Are you happy if the function replaces the property (breaking change) or do you want to keep both?

Let's keep both but deprecate the property if possible. We keep all of our deprecations here for now:

https://github.com/pointfreeco/swift-composable-architecture/blob/a03611d45da994a706911ca024fa441b148e79cb/Sources/ComposableArchitecture/Internal/Deprecations.swift

> * What other places would you like to see this change? `forEach` maybe?

Definitely on `forEach`! Ideally any higher-order reducer that can assert.

stephencelis commented on Sources/ComposableArchitecture/Reducer.swift:
```diff
@@ -189,6 +189,59 @@ public struct Reducer<State, Action, Environment> {
+            * This action was sent to the store while state was "nil". Make sure that actions for \
+            this reducer can only be sent to a view store when state is non-"nil". In SwiftUI \
+            applications, use "IfLetStore".
+            """
```
I think rather than embed the file/line in the message above, they'll be automatically added if you pass them along here:

```suggestion
            """,
            file: file,
            line: line
```

alexito4 replied:
Much better thanks.

stephencelis commented on Sources/ComposableArchitecture/Reducer.swift:
```diff
@@ -189,6 +189,59 @@ public struct Reducer<State, Action, Environment> {
+    ///   store on non-optional state.
+    /// - See also: `Store.ifLet`, a UIKit helper for doing imperative work with a store on optional
+    ///   state.
+    public func optional(_ file: String = #file, _ line: Int = #line) -> Reducer<State?, Action, Environment> {
```
I think `assertionFailure` parameters are `StaticString` and `Uint`...

```suggestion
    public func optional(_ file: StaticString = #file, _ line: UInt = #line) -> Reducer<State?, Action, Environment> {
```

alexito4 replied:
Yes you're totally right
![Screenshot 2020-07-08 at 19 22 56](https://user-images.githubusercontent.com/750912/86950421-74d6cc80-c150-11ea-8ddb-7d470a23e006.png)

stephencelis commented on Sources/ComposableArchitecture/Reducer.swift:
```diff
@@ -189,6 +189,61 @@ public struct Reducer<State, Action, Environment> {
+        guard state != nil else {
+          assertionFailure(
+            """
+            "\(debugCaseOutput(action))" was received by an optional reducer at \(file):\(line) \
```
Assuming `file` and `line` make it into the console in a reasonable manner wanna change this messaging to be the same as the existing one?

alexito4 replied:
Yes that's what I was changing now but then I realised I don't see any difference passing file/line to the assert 🤔 I was expecting the error line to stop at the point of origin like what XCTest assertions do. 
![Screenshot 2020-07-08 at 19 30 35](https://user-images.githubusercontent.com/750912/86951255-8e2c4880-c151-11ea-8d0f-fc99b71a3144.png)

stephencelis replied:
Isn't it showing up at the very end here?

<img width="764" alt="image" src="https://user-images.githubusercontent.com/658/86956525-34af2400-c127-11ea-88f2-e28acdf85c62.png">

stephencelis replied:
And yeah it'd be nice for Xcode to trap on those lines instead, too :( Maybe feedback should be filed.

alexito4 replied:
You're right! I thought that was part of the message somehow. All good then 👍

alexito4 commented:
Thanks for the feedback.

- Made the optional property deprecated:
![Screenshot 2020-07-08 at 19 21 18](https://user-images.githubusercontent.com/750912/86950285-435e0100-c150-11ea-834a-fc3f9316ec9d.png)
- Updated examples to use the new optional function.
- Added file/line to all forEach variations. 
- I haven't found other high order reducers that can fail, not on the lib or in any example.
- Since Xcode doesn't seem to show up the file/line anywhere https://github.com/pointfreeco/swift-composable-architecture/pull/218#discussion_r451712447 should we add the info in the message itself?
- Should I mark the deprecation as unavailable as per #219 or is gonna be to early for that?
- swift-format run locally ^^

stephencelis commented:
> * Should I mark the deprecation as unavailable as per #219 or is gonna be to early for that?

We can wait a few versions to change this "deprecated" to "unavailable" so that folks have a chance to more slowly migrate to the new APIs rather than force them to all at once.

alexito4 commented:
All good then. Ping me if you need any further changes 👍
